README.txt
= net_http_persistent

* http://seattlerb.rubyforge.org/net-http-persistent

== DESCRIPTION:

Persistent connections using Net::HTTP plus a speed fix for 1.8.  It's
thread-safe too!

== FORK DESCRIPTION

This is an experimental branch that implements a connection pool of 
Net::HTTP objects instead of a connection/thread.  C/T is fine if
you're only using your http threads to make connections but if you 
use them in child threads then I suspect you will have a thread memory
leak.  Also, I want to see if I get less connection resets if the
most recently used connection is always returned.

Also added a :force_retry option that if set to true will retry POST
requests as well as idempotent requests.

This branch is currently incompatible with the master branch in the 
following ways:

* It doesn't allow you to recreate the Net::HTTP::Persistent object
  on the fly.  This is possible in the master version since all the 
  data is kept in thread local storage.  For this version, you should
  probably create a class instance of the object and use that in your
  instance methods.

* It uses a hash in the initialize method.  This was easier for me
  as I use a HashWithIndifferentAccess created from a YAML file to 
  define my options.  This should probably be modified to check the
  arguments to achieve backwards compatibility.

* The method shutdown is unimplemented as I wasn't sure how I should
  implement it and I don't need it as I do a graceful shutdown from 
  nginx to finish up my connections.

For connection issues, I completely recreate a new Net::HTTP instance.
I was running into an issue which I suspect is a JRuby bug where an
SSL connection that times out would leave the ssl context in a frozen
state which would then make that connection unusable so each time that
thread handled a connection a 500 error with the exception "TypeError:
can't modify frozen".  I think Joseph West's fork resolves this issue
but I'm paranoid so I recreate the object.

Compatibility with the master version could probably be achieved by
creating a Strategy wrapper class for GenePool and a separate strategy
class with the connection/thread implementation.

== FEATURES/PROBLEMS:

* Supports SSL
* Thread-safe
* Pure ruby
* Timeout-less speed boost for 1.8 (by Aaron Patterson)

== INSTALL:

  gem install bpardee-net-http-persistent

== EXAMPLE USAGE:

  class MyHttpClient
    @@http ||= Net::HTTP::Persistent.new(
      :name         => 'MyHttpClient',
      :logger       => Rails.logger,
      :pool_size    => 10,
      :warn_timeout => 0.25,
      :force_retry  => true
    )

    def send_get_message
      uri = URI.parse('https://www.example.com/echo/foo')
      response = @@http.request(uri)
      ... Handle response as you would a normal Net::HTTPResponse ...
    end

    def send_post_message
      uri = URI.parse('https://www.example.com/echo/foo')
      request = Net::HTTP::Post.new(uri.request_uri)
      ... Modify request as needed ...
      response = @@http.request(uri, request)
      ... Handle response as you would a normal Net::HTTPResponse ...
    end
  end

== LICENSE:

(The MIT License)

Copyright (c) 2010 Eric Hodel, Aaron Patterson

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


History.txt
=== 1.0.0 / 2010-09-08

* Major Enhancements
  * Fork of drbrain/net-http-persistent 1.3 prerelease
  * Uses a connection pool instead of connection/thread
  * force_retry option added to retry on POST as well as idempotent requests


test/test_net_http_persistent.rb
require 'rubygems'
require 'minitest/autorun'
require 'net/http/persistent'
require 'openssl'
require 'stringio'
require 'ostruct'
require 'logger'

CMD_SUCCESS      = 'success'
CMD_SLEEP        = 'sleep'
CMD_BAD_RESPONSE = 'bad_response'
CMD_EOF_ERROR    = 'eof_error'
CMD_CONNRESET    = 'connreset'
CMD_ECHO         = 'echo'

PASS = 'pass'
FAIL = 'fail'

DUMMY_OPEN_TIMEOUT_FOR_HOSTDOWN    = 9000
DUMMY_OPEN_TIMEOUT_FOR_CONNREFUSED = 9001

$debug = false

class Net::HTTP
  def connect
    raise Errno::EHOSTDOWN    if open_timeout == DUMMY_OPEN_TIMEOUT_FOR_HOSTDOWN
    raise Errno::ECONNREFUSED if open_timeout == DUMMY_OPEN_TIMEOUT_FOR_CONNREFUSED
  end

  def successful_response
    r = Net::HTTPResponse.allocate
    def r.http_version() '1.1' end
    def r.read_body() :read_body end
    yield r if block_given?
    r
  end

  def request(req, &block)
    $count += 1
    puts "path=#{req.path} count=#{$count}" if $debug
    args = req.path[1..-1].split('/')
    cmd = args.shift
    i = $count % args.size if args.size > 0
    puts "i=#{i}" if $debug
    if cmd == CMD_ECHO
      res = successful_response(&block)
      eval "def res.body() \"#{req.body}\" end"
      return res
    elsif cmd == CMD_SUCCESS || args[i] == PASS
      return successful_response(&block)
    end
    case cmd
    when CMD_SLEEP
      sleep args[i].to_i
      return successful_response(&block)
    when CMD_BAD_RESPONSE
      raise Net::HTTPBadResponse.new('Dummy bad response') 
    when CMD_EOF_ERROR
      raise EOFError.new('Dummy EOF error') 
    when CMD_CONNRESET
      raise Errno::ECONNRESET
    else
      return successful_response(&block)
    end 
  end
end

class Net::HTTP::Persistent
  attr_reader :pool_hash
  
  # Make private methods public
  send(:public, *(self.private_instance_methods - Object.private_instance_methods))
end

class TestNetHttpPersistent < MiniTest::Unit::TestCase

  def uri_for(*args)
    URI.parse("http://example.com/#{args.join('/')}")
  end
  
  def request_command(req, *args)
    puts "uri=#{uri_for(args)}" if $debug
    @http.request(uri_for(args), req)
  end
  
  def http_and_io(options={})
    io = StringIO.new
    logger = Logger.new(io)
    logger.level = Logger::INFO
    default_options = {:name => 'TestNetHTTPPersistent', :logger => logger, :pool_size => 1}
    http = Net::HTTP::Persistent.new(default_options.merge(options))
    [http, io]
  end

  def setup
    $count = -1
    @http, @io = http_and_io
    @uri  = uri_for CMD_SUCCESS

    ENV.delete 'http_proxy'
    ENV.delete 'HTTP_PROXY'
    ENV.delete 'http_proxy_user'
    ENV.delete 'HTTP_PROXY_USER'
    ENV.delete 'http_proxy_pass'
    ENV.delete 'HTTP_PROXY_PASS'
  end

  def teardown
  end

  def test_initialize
    assert_nil @http.proxy_uri
  end

  def test_initialize_name
    http = Net::HTTP::Persistent.new(:name => 'name')
    assert_equal 'name', http.name
  end

  def test_initialize_env
    ENV['HTTP_PROXY'] = 'proxy.example'
    http = Net::HTTP::Persistent.new(:proxy => :ENV)

    assert_equal URI.parse('http://proxy.example'), http.proxy_uri
  end

  def test_initialize_uri
    proxy_uri = URI.parse 'http://proxy.example'

    http = Net::HTTP::Persistent.new(:proxy => proxy_uri)

    assert_equal proxy_uri, http.proxy_uri
  end

  def test_connection
    @http.open_timeout = 123
    @http.read_timeout = 321
    pool = @http.pool_for @uri
    assert_same pool, @http.pool_hash['example.com:80']
    pool.with_connection do |c|
      assert c.started?
      refute c.proxy?

      assert_equal 123, c.open_timeout
      assert_equal 321, c.read_timeout
      
      assert_equal 'example.com', c.address
      assert_equal 80, c.port
    end
  end

  def test_connection_for_cached
    c1, c2 = nil, nil
    pool = @http.pool_for @uri
    assert_same pool, @http.pool_hash['example.com:80']
    pool.with_connection do |c|
      c1 = c
      assert c.started?
    end
    pool.with_connection do |c|
      c2 = c
      assert c.started?
    end
    assert_same c1,c2
  end

  def test_connection_for_debug_output
    io = StringIO.new
    @http.debug_output = io

    pool = @http.pool_for @uri
    assert_same pool, @http.pool_hash['example.com:80']
    pool.with_connection do |c|
      assert c.started?
      assert_equal io, c.instance_variable_get(:@debug_output)
      assert_equal 'example.com', c.address
      assert_equal 80, c.port
    end
  end

  def test_connection_for_hostdown
    @http.open_timeout = DUMMY_OPEN_TIMEOUT_FOR_HOSTDOWN
    e = assert_raises Net::HTTP::Persistent::Error do
      request_command(nil, CMD_SUCCESS)
    end

    assert_match %r%host down%, e.message
  end

  def test_connection_for_connrefused
    @http.open_timeout = DUMMY_OPEN_TIMEOUT_FOR_CONNREFUSED
    e = assert_raises Net::HTTP::Persistent::Error do
      request_command(nil, CMD_SUCCESS)
    end

    assert_match %r%connection refused%, e.message
  end

  def test_connection_for_proxy
    uri = URI.parse 'http://proxy.example'
    uri.user     = 'johndoe'
    uri.password = 'muffins'

    http = Net::HTTP::Persistent.new(:proxy => uri)
    pool = http.pool_for(@uri)
    assert_same pool, http.pool_hash['example.com:80:proxy.example:80:johndoe:muffins']
    pool.with_connection do |c|
      assert c.started?
      assert c.proxy?
    end
  end

  def test_error_message
    6.times do
      request_command nil, CMD_EOF_ERROR, PASS, PASS, PASS, PASS, FAIL, PASS, PASS
    end

    assert_match "after 5 requests on", @io.string
  end

  def test_escape
    assert_nil @http.escape nil

    assert_equal '%20', @http.escape(' ')
  end

  def test_finish
    c = Object.new
    def c.finish; @finished = true end
    def c.finished?; @finished end
    def c.start; @started = true end
    def c.started?; @started end

    @http.finish c

    refute c.started?
    assert c.finished?
  end

  def test_finish_io_error
    c = Object.new
    def c.finish; @finished = true; raise IOError end
    def c.finished?; @finished end
    def c.start; @started = true end
    def c.started?; @started end

    @http.finish c

    refute c.started?
    assert c.finished?
  end
    
  def test_http_version
    assert_nil @http.http_version @uri
  
    request_command nil, CMD_SUCCESS
  
    assert_equal '1.1', @http.http_version(@uri)
  end
  
  def test_idempotent_eh
    assert @http.idempotent? Net::HTTP::Delete.new '/'
    assert @http.idempotent? Net::HTTP::Get.new '/'
    assert @http.idempotent? Net::HTTP::Head.new '/'
    assert @http.idempotent? Net::HTTP::Options.new '/'
    assert @http.idempotent? Net::HTTP::Put.new '/'
    assert @http.idempotent? Net::HTTP::Trace.new '/'
  
    refute @http.idempotent? Net::HTTP::Post.new '/'
  end
  
  def test_normalize_uri
    assert_equal 'http://example',  @http.normalize_uri('example')
    assert_equal 'http://example',  @http.normalize_uri('http://example')
    assert_equal 'https://example', @http.normalize_uri('https://example')
  end
  
  def test_proxy_from_env
    ENV['HTTP_PROXY']      = 'proxy.example'
    ENV['HTTP_PROXY_USER'] = 'johndoe'
    ENV['HTTP_PROXY_PASS'] = 'muffins'
  
    uri = @http.proxy_from_env
  
    expected = URI.parse 'http://proxy.example'
    expected.user     = 'johndoe'
    expected.password = 'muffins'
  
    assert_equal expected, uri
  end
  
  def test_proxy_from_env_lower
    ENV['http_proxy']      = 'proxy.example'
    ENV['http_proxy_user'] = 'johndoe'
    ENV['http_proxy_pass'] = 'muffins'
  
    uri = @http.proxy_from_env
  
    expected = URI.parse 'http://proxy.example'
    expected.user     = 'johndoe'
    expected.password = 'muffins'
  
    assert_equal expected, uri
  end
  
  def test_proxy_from_env_nil
    uri = @http.proxy_from_env
  
    assert_nil uri
  
    ENV['HTTP_PROXY'] = ''
  
    uri = @http.proxy_from_env
  
    assert_nil uri
  end
  
  def test_remove
    http, io = http_and_io(:pool_size => 3)
    request_command(nil, CMD_SUCCESS)
    pool = http.pool_for(@uri)
    2.times do
      conns = []
      pool.with_connection do |c1|
        pool.with_connection do |c2|
          conns << c2
          pool.with_connection do |c3|
            conns << c3
            begin
              Timeout.timeout(2) do
                pool.with_connection { |c4| }
                flunk('should NOT have been able to get 4th connection')
              end
            rescue  Timeout::Error => e
              pass('successfully failed to get a connection')
            end
            http.remove(pool, c1)
            Timeout.timeout(1) do
              begin
                pool.with_connection do |c4|
                  conns << c4
                end
              rescue  Timeout::Error => e
                flunk('should have been able to get 4th connection')
              end
            end
          end
        end
      end
      pool.with_connection do |c1|
        pool.with_connection do |c2|
          pool.with_connection do |c3|
            assert_equal conns, [c1,c2,c3]
          end
        end
      end
      # Do it a 2nd time with finish returning an IOError
      c1 = conns[0]
      def c1.finish
        super
        raise IOError
      end
    end
  end

  def test_renew
    http, io = http_and_io(:pool_size => 3)
    request_command(nil, CMD_SUCCESS)
    pool = http.pool_for(@uri)
    2.times do
      conns = []
      pool.with_connection do |c1|
        pool.with_connection do |c2|
          conns << c2
          pool.with_connection do |c3|
            conns << c3
            new_c1 = http.renew(pool, c1)
            refute_equal(c1, new_c1)
            conns.unshift(new_c1)
          end
        end
      end
      pool.with_connection do |c1|
        pool.with_connection do |c2|
          pool.with_connection do |c3|
            assert_equal conns, [c1,c2,c3]
          end
        end
      end
      # Do it a 2nd time with finish returning an IOError
      c1 = conns[0]
      def c1.finish
        super
        raise IOError
      end
    end
  end
  
  def test_renew_with_exception
    http, io = http_and_io(:pool_size => 2)
    pool = http.pool_for(@uri)
    [[DUMMY_OPEN_TIMEOUT_FOR_HOSTDOWN, %r%host down%], [DUMMY_OPEN_TIMEOUT_FOR_CONNREFUSED, %r%connection refused%]].each do |pair|
      dummy_open_timeout = pair.first
      error_message = pair.last
      pool.with_connection do |c|
        old_c = c
        http.open_timeout = dummy_open_timeout
        e = assert_raises Net::HTTP::Persistent::Error do
          new_c = http.renew pool, c
        end
        assert_match error_message, e.message
      
        # Make sure our pool is still in good shape
        http.open_timeout = 5   # Any valid timeout will do
        pool.with_connection do |c1|
          refute_equal old_c, c1
          pool.with_connection do |c2|
            refute_equal old_c, c2
          end
        end
      end
    end
  end
   
  def test_request
    @http.headers['user-agent'] = 'test ua'
    req = Net::HTTP::Get.new(@uri.request_uri)
    res = @http.request(@uri, req)
  
    assert_kind_of Net::HTTPResponse, res
  
    assert_kind_of Net::HTTP::Get, req
    assert_equal @uri.path,    req.path
    assert_equal 'keep-alive', req['connection']
    assert_equal '30',         req['keep-alive']
    assert_match %r%test ua%,  req['user-agent']
  end
  
  def test_request_block
    @http.headers['user-agent'] = 'test ua'
    body = nil
    
    req = Net::HTTP::Get.new(@uri.request_uri)
    res = @http.request(@uri, req) do |r|
      body = r.read_body
    end
  
    assert_kind_of Net::HTTPResponse, res
    refute_nil body
  
    assert_kind_of Net::HTTP::Get, req
    assert_equal @uri.path,    req.path
    assert_equal 'keep-alive', req['connection']
    assert_equal '30',         req['keep-alive']
    assert_match %r%test ua%,  req['user-agent']
  end

  def test_request_bad_response
    e = assert_raises Net::HTTP::Persistent::Error do
      request_command nil, CMD_BAD_RESPONSE, FAIL, FAIL
    end
    assert_match %r%too many bad responses%, e.message
    assert_match %r%Renewing connection because of bad response%, @io.string
    assert_match %r%Removing connection because of too many bad responses%, @io.string

    res = request_command nil, CMD_SUCCESS
    assert_kind_of Net::HTTPResponse, res
  end

  def test_request_connreset
    e = assert_raises Net::HTTP::Persistent::Error do
      request_command nil, CMD_CONNRESET, FAIL, FAIL
    end
  
    assert_match %r%too many connection resets%, e.message
    assert_match %r%Renewing connection %, @io.string
    assert_match %r%Removing connection %, @io.string

    res = request_command nil, CMD_SUCCESS
    assert_kind_of Net::HTTPResponse, res
  end
  
  def test_request_bad_response_retry
    res = request_command nil, CMD_BAD_RESPONSE, FAIL, PASS
    assert_match %r%Renewing connection because of bad response%, @io.string
    assert_kind_of Net::HTTPResponse, res
  end
 
  def test_request_connreset_retry
    res = request_command nil, CMD_CONNRESET, FAIL, PASS
    assert_match %r%Renewing connection %, @io.string
    assert_kind_of Net::HTTPResponse, res
  end

  def test_request_bad_response_post
    uri = uri_for CMD_BAD_RESPONSE, FAIL, PASS
    post = Net::HTTP::Post.new(uri.request_uri)
    e = assert_raises Net::HTTP::Persistent::Error do
      request_command post, CMD_BAD_RESPONSE, FAIL, PASS
    end
    assert_match %r%too many bad responses%, e.message
    assert_match %r%Removing connection because of too many bad responses%, @io.string

    res = request_command nil, CMD_SUCCESS
    assert_kind_of Net::HTTPResponse, res
  end

  
  def test_request_connreset_post
    uri = uri_for CMD_CONNRESET, FAIL, PASS
    post = Net::HTTP::Post.new(uri.request_uri)
    e = assert_raises Net::HTTP::Persistent::Error do
      request_command post, CMD_CONNRESET, FAIL, PASS
    end
    assert_match %r%too many connection resets%, e.message
    assert_match %r%Removing connection %, @io.string

    res = request_command nil, CMD_SUCCESS
    assert_kind_of Net::HTTPResponse, res
  end
  
  def test_request_bad_response_post_force_retry
    @http.force_retry = true
    uri = uri_for CMD_BAD_RESPONSE, FAIL, PASS
    post = Net::HTTP::Post.new(uri.request_uri)
    res = request_command post, CMD_BAD_RESPONSE, FAIL, PASS
    assert_match %r%Renewing connection because of bad response%, @io.string
    assert_kind_of Net::HTTPResponse, res
  end
    
  def test_request_connreset_post_force_retry
    @http.force_retry = true
    uri = uri_for CMD_CONNRESET, FAIL, PASS
    post = Net::HTTP::Post.new(uri.request_uri)
    res = request_command post, CMD_CONNRESET, FAIL, PASS
    assert_match %r%Renewing connection %, @io.string
    assert_kind_of Net::HTTPResponse, res
  end
  
  def test_request_post
    uri = uri_for CMD_ECHO
    post = Net::HTTP::Post.new(uri.request_uri)
    post.body = 'hello Net::HTTP::Persistent'
    res = request_command post, CMD_ECHO
    assert_kind_of Net::HTTPResponse, res
    assert_equal post.body, res.body
  end
  
  # def test_shutdown
  #   c = connection
  #   cs = conns
  #   rs = reqs
  # 
  #   orig = @http
  #   @http = Net::HTTP::Persistent.new 'name'
  #   c2 = connection
  # 
  #   orig.shutdown
  # 
  #   assert c.finished?
  #   refute c2.finished?
  # 
  #   refute_same cs, conns
  #   refute_same rs, reqs
  # end
  # 
  # def test_shutdown_not_started
  #   c = Object.new
  #   def c.finish() raise IOError end
  # 
  #   conns["#{@uri.host}:#{@uri.port}"] = c
  # 
  #   @http.shutdown
  # 
  #   assert_nil Thread.current[@http.connection_key]
  #   assert_nil Thread.current[@http.request_key]
  # end
  # 
  # def test_shutdown_no_connections
  #   @http.shutdown
  # 
  #   assert_nil Thread.current[@http.connection_key]
  #   assert_nil Thread.current[@http.request_key]
  # end
  
  def test_ssl
    @http.verify_callback = :callback
    c = Net::HTTP.new 'localhost', 80
  
    @http.ssl c
  
    assert c.use_ssl?
    assert_equal OpenSSL::SSL::VERIFY_NONE, c.verify_mode
    assert_nil c.verify_callback
  end
  
  def test_ssl_ca_file
    @http.ca_file = 'ca_file'
    @http.verify_callback = :callback
    c = Net::HTTP.new 'localhost', 80
  
    @http.ssl c
  
    assert c.use_ssl?
    assert_equal OpenSSL::SSL::VERIFY_PEER, c.verify_mode
    assert_equal :callback, c.verify_callback
  end
  
  def test_ssl_certificate
    @http.certificate = :cert
    @http.private_key = :key
    c = Net::HTTP.new 'localhost', 80
  
    @http.ssl c
  
    assert c.use_ssl?
    assert_equal :cert, c.cert
    assert_equal :key,  c.key
  end
  
  def test_ssl_verify_mode
    @http.verify_mode = OpenSSL::SSL::VERIFY_NONE
    c = Net::HTTP.new 'localhost', 80
  
    @http.ssl c
  
    assert c.use_ssl?
    assert_equal OpenSSL::SSL::VERIFY_NONE, c.verify_mode
  end
end



lib/net/http/persistent.rb
require 'net/http'
require 'net/http/faster'
require 'uri'
require 'gene_pool'

##
# Persistent connections for Net::HTTP
#
# Net::HTTP::Persistent maintains persistent connections across all the
# servers you wish to talk to.  For each host:port you communicate with a
# single persistent connection is created.
#
# Multiple Net::HTTP::Persistent objects will share the same set of
# connections which will be checked out of a pool.
#
# You can shut down the HTTP connections when done by calling #shutdown.  You
# should name your Net::HTTP::Persistent object if you intend to call this
# method.
#
# Example:
#
#   uri = URI.parse 'http://example.com/awesome/web/service'
#   http = Net::HTTP::Persistent.new
#   stuff = http.request uri # performs a GET
#
#   # perform a POST
#   post_uri = uri + 'create'
#   post = Net::HTTP::Post.new post_uri.path
#   post.set_form_data 'some' => 'cool data'
#   http.request post_uri, post # URI is always required

class Net::HTTP::Persistent

  ##
  # The version of Net::HTTP::Persistent use are using

  VERSION = '1.0.0'

  ##
  # Error class for errors raised by Net::HTTP::Persistent.  Various
  # SystemCallErrors are re-raised with a human-readable message under this
  # class.

  class Error < StandardError; end

  ##
  # An SSL certificate authority.  Setting this will set verify_mode to
  # VERIFY_PEER.

  attr_accessor :ca_file

  ##
  # This client's OpenSSL::X509::Certificate

  attr_accessor :certificate

  ##
  # Sends debug_output to this IO via Net::HTTP#set_debug_output.
  #
  # Never use this method in production code, it causes a serious security
  # hole.

  attr_accessor :debug_output

  ##
  # Retry even for non-idempotent (POST) requests.

  attr_accessor :force_retry

  ##
  # Headers that are added to every request

  attr_accessor :headers

  ##
  # Maps host:port to an HTTP version.  This allows us to enable version
  # specific features.

  attr_reader :http_versions

  ##
  # The value sent in the Keep-Alive header.  Defaults to 30.  Not needed for
  # HTTP/1.1 servers.
  #
  # This may not work correctly for HTTP/1.0 servers
  #
  # This method may be removed in a future version as RFC 2616 does not
  # require this header.

  attr_accessor :keep_alive

  ##
  # Logger for message logging.

  attr_accessor :logger

  ##
  # A name for this connection.  Allows you to keep your connections apart
  # from everybody else's.

  attr_reader :name

  ##
  # Seconds to wait until a connection is opened.  See Net::HTTP#open_timeout

  attr_accessor :open_timeout

  ##
  # The maximum size of the connection pool

  attr_reader :pool_size

  ##
  # This client's SSL private key

  attr_accessor :private_key

  ##
  # The URL through which requests will be proxied

  attr_reader :proxy_uri

  ##
  # Seconds to wait until reading one block.  See Net::HTTP#read_timeout

  attr_accessor :read_timeout

  ##
  # SSL verification callback.  Used when ca_file is set.

  attr_accessor :verify_callback

  ##
  # HTTPS verify mode.  Defaults to OpenSSL::SSL::VERIFY_NONE which ignores
  # certificate problems.
  #
  # You can use +verify_mode+ to override any default values.

  attr_accessor :verify_mode

  ##
  # The threshold in seconds for checking out a connection at which a warning 
  # will be logged via the logger

  attr_accessor :warn_timeout

  ##
  # Creates a new Net::HTTP::Persistent.
  #
  # Set +name+ to keep your connections apart from everybody else's.  Not
  # required currently, but highly recommended.  Your library name should be
  # good enough.  This parameter will be required in a future version.
  #
  # +proxy+ may be set to a URI::HTTP or :ENV to pick up proxy options from
  # the environment.  See proxy_from_env for details.
  #
  # In order to use a URI for the proxy you'll need to do some extra work
  # beyond URI.parse:
  #
  #   proxy = URI.parse 'http://proxy.example'
  #   proxy.user     = 'AzureDiamond'
  #   proxy.password = 'hunter2'

  def initialize(options={})
    @name = options[:name] || 'Net::HTTP::Persistent'
    proxy = options[:proxy]

    @proxy_uri = case proxy
                 when :ENV      then proxy_from_env
                 when URI::HTTP then proxy
                 when nil       then # ignore
                 else raise ArgumentError, 'proxy must be :ENV or a URI::HTTP'
                 end

    if @proxy_uri then
      @proxy_args = [
        @proxy_uri.host,
        @proxy_uri.port,
        @proxy_uri.user,
        @proxy_uri.password,
      ]

      @proxy_connection_id = [nil, *@proxy_args].join ':'
    end

    @ca_file         = options[:ca_file]
    @certificate     = options[:certificate]
    @debug_output    = options[:debug_output]
    @force_retry     = options[:force_retry]
    @headers         = options[:header]          || {}
    @http_versions   = {}
    @keep_alive      = options[:keep_alive]      || 30
    @logger          = options[:logger]
    @open_timeout    = options[:open_timeout]
    @pool_size       = options[:pool_size]       || 1
    @private_key     = options[:private_key]
    @read_timeout    = options[:read_timeout]
    @verify_callback = options[:verify_callback]
    @verify_mode     = options[:verify_mode]
    @warn_timeout    = options[:warn_timeout]    || 0.5
    
    # Hash containing connection pools based on key of host:port
    @pool_hash = {}
    
    # Hash containing the request counts based on the connection
    @count_hash = Hash.new(0)
  end

  ##
  # Makes a request on +uri+.  If +req+ is nil a Net::HTTP::Get is performed
  # against +uri+.
  #
  # If a block is passed #request behaves like Net::HTTP#request (the body of
  # the response will not have been read).
  #
  # +req+ must be a Net::HTTPRequest subclass (see Net::HTTP for a list).
  #
  # If there is an error and the request is idempontent according to RFC 2616
  # it will be retried automatically.

  def request uri, req = nil, &block
    retried      = false
    bad_response = false

    req = Net::HTTP::Get.new uri.request_uri unless req

    headers.each do |pair|
      req.add_field(*pair)
    end

    req.add_field 'Connection', 'keep-alive'
    req.add_field 'Keep-Alive', @keep_alive

    pool = pool_for uri
    pool.with_connection do |connection|
      begin
        count = @count_hash[connection.object_id] += 1
        response = connection.request req, &block
        @http_versions["#{uri.host}:#{uri.port}"] ||= response.http_version
        return response

      rescue  Timeout::Error => e
        due_to = "(due to #{e.message} - #{e.class})"
        message = error_message connection
        @logger.info "#{name}: Removing connection #{due_to} #{message}" if @logger
        remove pool, connection
        raise
        
      rescue Net::HTTPBadResponse => e
        message = error_message connection
        if bad_response or not (idempotent? req or @force_retry)
          @logger.info "#{name}: Removing connection because of too many bad responses #{message}" if @logger
          remove pool, connection
          raise Error, "too many bad responses #{message}"
        else
          bad_response = true
          @logger.info "#{name}: Renewing connection because of bad response #{message}" if @logger
          connection = renew pool, connection
          retry
        end

      rescue IOError, EOFError, Errno::ECONNABORTED, Errno::ECONNRESET, Errno::EPIPE => e
        due_to = "(due to #{e.message} - #{e.class})"
        message = error_message connection
        if retried or not (idempotent? req or @force_retry)
          @logger.info "#{name}: Removing connection #{due_to} #{message}" if @logger
          remove pool, connection
          raise Error, "too many connection resets #{due_to} #{message}"
        else
          retried = true
          @logger.info "#{name}: Renewing connection #{due_to} #{message}" if @logger
          connection = renew pool, connection
          retry
        end
      end
    end
  end

  ##
  # Returns the HTTP protocol version for +uri+

  def http_version uri
    @http_versions["#{uri.host}:#{uri.port}"]
  end

  ##
  # Shuts down all connections.

  def shutdown
    raise 'Shutdown not implemented'
    # TBD - need to think about this one
    @count_hash = nil
  end

  #######
  private
  #######

  ##
  # Returns an error message containing the number of requests performed on
  # this connection

  def error_message connection
    requests = @count_hash[connection.object_id] || 0
    "after #{requests} requests on #{connection.object_id}"
  end

  ##
  # URI::escape wrapper

  def escape str
    URI.escape str if str
  end

  ##
  # Finishes the Net::HTTP +connection+

  def finish connection
    @count_hash.delete(connection.object_id)
    connection.finish
  rescue IOError
  end

  ##
  # Is +req+ idempotent according to RFC 2616?

  def idempotent? req
    case req
    when Net::HTTP::Delete, Net::HTTP::Get, Net::HTTP::Head,
         Net::HTTP::Options, Net::HTTP::Put, Net::HTTP::Trace then
      true
    end
  end

  ##
  # Adds "http://" to the String +uri+ if it is missing.

  def normalize_uri uri
    (uri =~ /^https?:/) ? uri : "http://#{uri}"
  end

  ##
  # Get the connection pool associated with this +uri+
  def pool_for uri
    net_http_args = [uri.host, uri.port]
    connection_id = net_http_args.join ':'

    if @proxy_uri then
      connection_id << @proxy_connection_id
      net_http_args.concat @proxy_args
    end
    @pool_hash[connection_id] ||= GenePool.new(:name         => name + '-' + connection_id,
                                               :pool_size    => @pool_size,
                                               :warn_timeout => @warn_timeout,
                                               :logger       => @logger) do
      begin
        connection = Net::HTTP.new(*net_http_args)
        connection.set_debug_output @debug_output if @debug_output
        connection.open_timeout = @open_timeout if @open_timeout
        connection.read_timeout = @read_timeout if @read_timeout

        ssl connection if uri.scheme == 'https'

        connection.start
        connection
      rescue Errno::ECONNREFUSED
        raise Error, "connection refused: #{connection.address}:#{connection.port}"
      rescue Errno::EHOSTDOWN
        raise Error, "host down: #{connection.address}:#{connection.port}"
      end
    end
  end

  ##
  # Creates a URI for an HTTP proxy server from ENV variables.
  #
  # If +HTTP_PROXY+ is set a proxy will be returned.
  #
  # If +HTTP_PROXY_USER+ or +HTTP_PROXY_PASS+ are set the URI is given the
  # indicated user and password unless HTTP_PROXY contains either of these in
  # the URI.
  #
  # For Windows users lowercase ENV variables are preferred over uppercase ENV
  # variables.

  def proxy_from_env
    env_proxy = ENV['http_proxy'] || ENV['HTTP_PROXY']

    return nil if env_proxy.nil? or env_proxy.empty?

    uri = URI.parse(normalize_uri(env_proxy))

    unless uri.user or uri.password then
      uri.user     = escape ENV['http_proxy_user'] || ENV['HTTP_PROXY_USER']
      uri.password = escape ENV['http_proxy_pass'] || ENV['HTTP_PROXY_PASS']
    end

    uri
  end

  ##
  # Finishes then removes the Net::HTTP +connection+

  def remove pool, connection
    finish connection
    pool.remove(connection)
  end

  ##
  # Finishes then renews the Net::HTTP +connection+.  It may be unnecessary 
  # to completely recreate the connection but connections that get timed out
  # in JRuby leave the ssl context in a frozen object state.

  def renew pool, connection
    finish connection
    connection = pool.renew(connection)
  end

  ##
  # Enables SSL on +connection+

  def ssl connection
    require 'net/https'
    connection.use_ssl = true

    # suppress warning but allow override
    connection.verify_mode = OpenSSL::SSL::VERIFY_NONE unless @verify_mode

    if @ca_file then
      connection.ca_file = @ca_file
      connection.verify_mode = OpenSSL::SSL::VERIFY_PEER
      connection.verify_callback = @verify_callback if @verify_callback
    end

    if @certificate and @private_key then
      connection.cert = @certificate
      connection.key  = @private_key
    end

    connection.verify_mode = @verify_mode if @verify_mode
  end
  
end

